### ISSUE TITLE
LSTM for MPS backend is broken

### ISSUE THREAD

**Issue by AngledLuffa** [pytorch/pytorch#90421]
### 🐛 Describe the bug

A bidirectional LSTM using the MPS backend gives bad results, with or without batch_first, regardless of the number of layers.

There was an existing bug report which addressed one aspect of this problem, but it was closed before the problem was fully resolved.  There are still several cases in which an LSTM on MPS is broken.

https://github.com/pytorch/pytorch/issues/80306

Here is a short script which demonstrates the issue:

```
import torch

def test_lstm(num_layers, bidirectional, batch_first):
    torch.manual_seed(1234)
    lstm = torch.nn.LSTM(5, 5, num_layers=num_layers, bidirectional=bidirectional, batch_first=batch_first)
    lstm.eval()
    inp = torch.randn(4, 1, 5)
    if batch_first:
        inp = inp.transpose(0, 1)
    print("num_layers {} bidirectional {} batch_first {}".format(num_layers, bidirectional, batch_first), end="  ")
    print(torch.linalg.norm(lstm(inp)[0]).item(), end="  ")
    print(torch.linalg.norm(lstm.to("mps")(inp.to("mps"))[0]).item())

test_lstm(2, True, True)   # bad
test_lstm(2, True, False)  # bad
test_lstm(1, True, True)   # bad

test_lstm(2, False, True)  # ok
test_lstm(1, False, True)  # ok
test_lstm(1, False, False) # ok
```

The results:

```
/Users/john/stanza/bar.py:12: UserWarning: The operator 'aten::linalg_vector_norm' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)
  print(torch.linalg.norm(lstm.to("mps")(inp.to("mps"))[0]).item())
num_layers 2 bidirectional True batch_first True  0.9131234288215637  2.432673692703247
num_layers 2 bidirectional True batch_first False  0.9131234288215637  2.432673692703247
num_layers 1 bidirectional True batch_first True  0.8858160972595215  2.428898334503174
num_layers 2 bidirectional False batch_first True  0.5294497013092041  0.5294497609138489
num_layers 1 bidirectional False batch_first True  0.6045955419540405  0.6045956611633301
num_layers 1 bidirectional False batch_first False  0.6045955419540405  0.6045956611633301
```

### Versions

Environment information:
```
Collecting environment information...
PyTorch version: 1.14.0.dev20221206
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.6 (default, Sep 26 2022, 11:37:49)  [Clang 14.0.0 (clang-1400.0.29.202)] (64-bit runtime)
Python platform: macOS-12.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==1.14.0.dev20221206
[pip3] torchaudio==0.14.0.dev20221206
[pip3] torchvision==0.15.0.dev20221206
[conda] Could not collect
```

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @kulinseth @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by qqaatw** [comment]
The cause has been identified, I'll send a PR after a prerequisite PR is merged.

**Comment by AngledLuffa** [comment]
Thank you, that would be excellent

**Comment by AngledLuffa** [comment]
Any progress?  It looks like the PR was rebased five days ago, so maybe that means it's almost there?

**Comment by qqaatw** [comment]
The PR involves some changes that slightly break the general behavior of PyTorch, so maybe they're discussing about it internally. If it doesn't matter too much, then the PR should be merged soon.

**Comment by AngledLuffa** [comment]
Awesome, thanks!

**Comment by AngledLuffa** [comment]
This is still broken.  It works for the one row of input version of an LSTM, but it has the same exact problem for multiple inputs in a batch.

```
import torch

def test_lstm(num_layers, bidirectional, batch_first):
    torch.manual_seed(1234)
    lstm = torch.nn.LSTM(5, 5, num_layers=num_layers, bidirectional=bidirectional, batch_first=batch_first)
    lstm.eval()
    inp = torch.randn(4, 2, 5)   # the only change I made was to make a batch of size 2
    if batch_first:
        inp = inp.transpose(0, 1)
    print("num_layers {} bidirectional {} batch_first {}".format(num_layers, bidirectional, batch_first), end="  ")
    print(torch.linalg.norm(lstm(inp)[0]).item(), end="  ")
    print(torch.linalg.norm(lstm.to("mps")(inp.to("mps"))[0]).item())

test_lstm(2, True, True)   # bad                                                                                                                                                     
test_lstm(2, True, False)  # bad                                                                                                                                                     
test_lstm(1, True, True)   # bad                                                                                                                                                     

test_lstm(2, False, True)  # ok                                                                                                                                                      
test_lstm(1, False, True)  # ok                                                                                                                                                      
test_lstm(1, False, False) # ok                                                                                                                                                      
```

results:

```
num_layers 2 bidirectional True batch_first True  1.3071484565734863  1.6624640226364136    # wrong
num_layers 2 bidirectional True batch_first False  1.3071484565734863  1.6624640226364136    # wrong
num_layers 1 bidirectional True batch_first True  1.241768479347229  2.869274616241455    # wrong
num_layers 2 bidirectional False batch_first True  0.747026801109314  0.7470268607139587
num_layers 1 bidirectional False batch_first True  1.085639238357544  1.0856393575668335
num_layers 1 bidirectional False batch_first False  1.085639238357544  1.0856393575668335
```

**Comment by qqaatw** [comment]
Hi, @AngledLuffa, 

here are the results on my side:

```
num_layers 2 bidirectional True batch_first True  1.3071484565734863  1.3071486949920654
num_layers 2 bidirectional True batch_first False  1.3071484565734863  1.3071486949920654
num_layers 1 bidirectional True batch_first True  1.241768479347229  1.241768717765808
num_layers 2 bidirectional False batch_first True  0.747026801109314  0.7470268607139587
num_layers 1 bidirectional False batch_first True  1.085639238357544  1.085639238357544
num_layers 1 bidirectional False batch_first False  1.085639238357544  1.085639238357544
```

Which macOS version and PyTorch version are you using now?

Note that the fix linked above has not been included in PyTorch v2.0.0. It is available on master/nightly only!

**Comment by AngledLuffa** [comment]
That might be it.  I was using 2.0.0, not the nightly.  I will try upgrading.

**Comment by AngledLuffa** [comment]
Yes, that works!  Thanks.

My impression was that the initial fix made it into 2.0.0, since it was working for an input of batch size 1, but perhaps there were two separate fixes or I otherwise missed some part of the fix.

**Comment by qqaatw** [comment]
The MPS LSTM implementation is separated into two paths, one that uses native MPS Graph API only supports macOS 13.0 onward since there are numerical correctness issues with the API in macOS 12; thus, we make multi-layer LSTM on macOS 12 fall back to LSTMCell, which is the other path and slower than the native one but at least workable!

**Comment by AngledLuffa** [comment]
I'm using 13.3.1, so that explanation makes a little suspicious that maybe it's falling back when it doesn't need to

**Comment by qqaatw** [comment]
Hmm that's strange, did you see a warning message like `Native multi-layer LSTM support in MPS available only on MacOS 13 onwards.` when you invoked LSTM?

**Comment by AngledLuffa** [comment]
I do not see such a warning.  My little sample program above warns about `linalg_vector_norm` falling back, but that's after the LSTM call.

To be clear, the problem I had was that a batch size of 2 was getting the wrong answer with pytorch 2.0, whereas with the nightly it is getting the right answer, while using MacOS 13.3.1.  Your statement that the fix for the native one (13+) is included in 2.0 makes me wonder if getting the wrong answer on version 2.0 means it was not using the native lstm for some reason.

**Comment by qqaatw** [comment]
OK I know what's going on. There is a fix https://github.com/pytorch/pytorch/pull/95563 for bi-directional LSTM that was made after 2.0 release branch was cut, and it was not cherry-picked into 2.0. So in 2.0 with bi-directional, you get a fallback LSTM (which has a numerical correctness issue at the moment). On master, that bi-directional problem was fixed by https://github.com/pytorch/pytorch/pull/95563 for the native one (1.13+). And the rest problem with multi-layer LSTM on macOS 12 falls back to LSTMCell with the numerical correctness issue fixed by the linked PR https://github.com/pytorch/pytorch/pull/90909.

So in short, you're now using the native one with all the issues fixed.

**Comment by AngledLuffa** [comment]
Sounds great, thanks!

**Comment by AngledLuffa** [comment]
fwiw I got stuck [here](https://github.com/pytorch/pytorch/issues/97552) now that the first issue has been resolved

**Comment by qqaatw** [comment]
> fwiw I got stuck [here](https://github.com/pytorch/pytorch/issues/97552) now that the first issue has been resolved

I'll take a look when I'm free. Since I'm still on macOS 12, not sure if that's reproducible on my side.
